File: _includes/code/howto/bq-compression.py
# THIS FILE NEEDS TESTS

# ==============================
# =====  DOWNLOAD DATA =====
# ==============================

# START DownloadData
import requests
import json

# Download the data
resp = requests.get(
    "https://raw.githubusercontent.com/weaviate-tutorials/intro-workshop/main/data/jeopardy_1k.json"
)

# Load the data so you can see what it is
data = json.loads(resp.text)

# Parse the JSON and preview it
print(type(data), len(data))
print(json.dumps(data[1], indent=2))

# END DownloadData

# ==============================
# =====  CONNECT =====
# ==============================

# START ConnectCode
import weaviate, os, json
import weaviate.classes as wvc

client = weaviate.connect_to_local(
    headers={
        "X-OpenAI-Api-Key": os.environ[
            "OPENAI_API_KEY"
        ]  # Replace with your OpenAI API key
    }
)

client.is_ready()

# END ConnectCode

# ==============================
# =====  EnableBQ =====
# ==============================

client.collections.delete("YourCollection")

# START EnableBQ
import weaviate.classes as wvc

client.collections.create(
    name="YourCollection",
    vectorizer_config=wvc.Configure.Vectorizer.text2vec_openai(),
    # highlight-start
    vector_index_config=wvc.Configure.VectorIndex.flat(),
    # highlight-end
    properties=[
        wvc.Property(name="title", data_type=wvc.DataType.TEXT),
        wvc.Property(name="body", data_type=wvc.DataType.TEXT),
    ]
)
# END EnableBQ

# ==============================
# =====  EnableBQ with Options =====
# ==============================

client.collections.delete("YourCollection")

# START BQWithOptions
import weaviate.classes as wvc

client.collections.create(
    name="YourCollection",
    vectorizer_config=wvc.Configure.Vectorizer.text2vec_openai(),
    # highlight-start
    vector_index_config=wvc.Configure.VectorIndex.flat(
        distance_metric=wvc.VectorDistance.COSINE,
        vector_cache_max_objects=1000000,
        quantizer=wvc.Configure.VectorIndex.Quantizer.bq()
    ),
    # highlight-end
    properties=[
        wvc.Property(name="title", data_type=wvc.DataType.TEXT),
        wvc.Property(name="body", data_type=wvc.DataType.TEXT),
    ]
)
# END BQWithOptions

File: _includes/code/howto/bq-compression.ts
import assert from 'assert';

// TODO - needs confirmation & test

// ================================
// ===== INSTANTIATION-COMMON =====
// ================================

// START DockerInstantiationExample
import weaviate, { WeaviateClient, ObjectsBatcher } from 'weaviate-ts-client';
import fetch from 'node-fetch';

const client: WeaviateClient = weaviate.client({
  scheme: 'http',
  host: 'localhost:8080',
  headers: { 'X-OpenAI-Api-Key': process.env.OPENAI_APIKEY }, // Replace with your inference API key
});
// END DockerInstantiationExample

if (client.schema.exists('YourCollection')) {
  console.log('Deleting class');
  await client.schema.classDeleter().withClassName('YourCollection').do();
}

// ==============================
// =====  EnableBQ =====
// ==============================

// START EnableBQ
async function enableBQ() {
  const classObj = {
    class: 'YourCollection',
    vectorizer: 'text2vec-openai', // Can be any vectorizer
    // highlight-start
    vectorIndexType: 'flat',
    vectorIndexConfig: {
      bq: {
        enabled: true,
      },
    },
    // highlight-end
    //  Remainder not shown
  };
  const res = await client.schema.classCreator().withClass(classObj).do();
  console.log(res);
}

await enableBQ();
// END EnableBQ

if (client.schema.exists('YourCollection')) {
  console.log('Deleting class');
  await client.schema.classDeleter().withClassName('YourCollection').do();
}

// ==============================
// =====  EnableBQ =====
// ==============================

// START BQWithOptions
async function bqWithOptions() {
  const classObj = {
    class: 'YourCollection',
    vectorizer: 'text2vec-openai', // Can be any vectorizer
    vectorIndexType: 'flat',
    vectorIndexConfig: {
      // highlight-start
      bq: {
        enabled: true,
        rescoreLimit: 200, // The minimum number of candidates to fetch before rescoring
        cache: true, // Default: false
      },
      // highlight-end
      vectorCacheMaxObjects: 100000, // Cache size (used if `cache` enabled)
    },
    //  Remainder not shown
  };
  const res = await client.schema.classCreator().withClass(classObj).do();
  console.log(res);
}

await bqWithOptions();
// END BQWithOptions
